drivers/soundwire/intel.c

```diff
@@ -1074,10 +1094,18 @@ intel_hw_free(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
 	if (!dma)
 		return -EIO;
 
-	ret = sdw_deprepare_stream(dma->stream);
-	if (ret) {
-		dev_err(dai->dev, "sdw_deprepare_stream: failed %d", ret);
-		return ret;
+	/*
+	 * All cpu dais belong to a stream. To ensure sdw_deprepare_stream
+	 * is called once per stream, we should call it only when
+	 * dai = first_cpu_dai.
+	 */
+	if (first_cpu_dai == dai) {
```

Comment:
I'm wondering in which order will this function be called for various CPU DAIs? Is it important to perform the freeing when the first one is processed or maybe the last one?

Reply:
> I'm wondering in which order will this function be called for various CPU DAIs? Is it important to perform the freeing when the first one is processed or maybe the last one?

The first cpu dai will be called first. So we will deprepare stream than remove master which is the right order.
ie. sdw_deprepare_stream() will be called before all sdw_stream_remove_master()
```
first cpu dai: deprepare stream -> remove master
other dais:                                         -> remove master
```

Reply:
Actually @lyakh has a valid point. We should add a comment that the sdw_stream_remove_master() will also lead to a stream state transitions when all master runtimes have been released. 

So the order is really
- transition to DEPREPARED for the first cpu-dai
- transition to RELEASED for the last cpu-dai

let's document this explicitly. I missed it to be honest. Thanks @lyakh for challenging us.

Reply:
Done. Thanks @plbossart @lyakh